### PR TITLE
Introduce & Refactor: `Libplanet.Common`

### PR DESCRIPTION
**Pull request by riemannulus**
For flexible extensibility between `Libplanet` and applications using `Libplanet`, common types and extensions used in `Libplanet` are moved to `Libplanet.Common`, a separate project.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -65,6 +73,11 @@ be compatible with this version.
     [[#3292]]
  -  Removed `IActionEvaluation.Logs` property. [[#3292]]
 
+
+### Backward-incompatible network protocol changes
+
+### Backward-incompatible storage format changes
+
```
Comment (greymistcube) [inline CHANGES.md:80]: Probably should be removed again. 🤔

**Comment by longfin** [comment]
I guess it seems slightly late but... `.Common` in the namespace seems a little bit verbose and can be a subjective term. how about `.Core`? (or, is it already dropped candidate? 😅)

File: Libplanet.Common/EnumerableExtensions.cs [hunk 2]
```diff
@@ -1,14 +1,14 @@
 #nullable disable
 using System;
 using System.Collections.Generic;
-#if !NETSTANDARD2_1
 using System.Diagnostics;
-#endif
 using System.Linq;
+#if !NETSTANDARD2_1
+#endif
```
Comment (greymistcube) [inline Libplanet.Common/EnumerableExtensions.cs:7]: 🤔

File: Libplanet.Common/HashDigest.cs [hunk 3]
```diff
@@ -1,3 +1,6 @@
+#if NETSTANDARD2_0
+using System.Runtime.CompilerServices;
+#endif
```
Comment (greymistcube) [inline Libplanet.Common/HashDigest.cs:3]: Is it okay not to be alphabetically sorted?

File: Libplanet.Common/Types/Blocks/BlockCommit.cs [hunk 4]
```diff
@@ -201,7 +201,7 @@ public override string ToString()
                 { "round", Round },
                 {
                     "votes",
-                    Votes.Select(
+                    Votes.ToArray().Select(
```
Comment (greymistcube) [inline Libplanet.Common/Types/Blocks/BlockCommit.cs:204]: Eh, why is this necessary?

File: Libplanet.Common/Types/Assets/FungibleAssetValue.cs [hunk 5]
```diff
@@ -112,7 +112,7 @@ public FungibleAssetValue(Currency currency, BigInteger majorUnit, BigInteger mi
         /// one of <c>1</c>, <c>0</c>, and <c>-1</c>, or <paramref name="majorUnit"/> or
         /// <paramref name="minorUnit"/> is negative.
         /// </exception>
-        /// <seealso cref="Assets.Currency.DecimalPlaces"/>
+        /// <seealso cref="Libplanet.Assets.Currency.DecimalPlaces"/>
```
Comment (greymistcube) [inline Libplanet.Common/Types/Assets/FungibleAssetValue.cs:115]: Namespace seems fishy. 🐟

File: Libplanet.Common/Types/Blocks/InvalidBlockSignatureException.cs [hunk 6]
```diff
@@ -28,7 +27,7 @@ public class InvalidBlockSignatureException : InvalidBlockException
             : base(
                 message +
                 (publicKey is { } pubKey ? $"\nPublic key: {pubKey}" : string.Empty) +
-                (invalidSignature is { } sig ? $"\nInvalid signature: {Hex(sig)}" : string.Empty)
+                (invalidSignature is { } sig ? $"\nInvalid signature: {sig}" : string.Empty)
```
Comment (greymistcube) [inline Libplanet.Common/Types/Blocks/InvalidBlockSignatureException.cs:30]: I think we need `Hex` here.

**Comment by riemannulus** [comment]
> how about .Core? (or, is it already dropped candidate?

It was in candidates. And yes, Common is not good enough. This PR is just suggestion so we need to discussion about new project name.

File: Libplanet.Common/EnumerableExtensions.cs (hunk 2 above)
Comment (riemannulus) [reply Libplanet.Common/EnumerableExtensions.cs:7]: Nah, I guess this is mistake.

File: Libplanet.Common/HashDigest.cs (hunk 3 above)
Comment (riemannulus) [reply Libplanet.Common/HashDigest.cs:3]: This is automatically sorted by rider. I'll get it back previous position.

File: Libplanet/Store/BaseStore.cs [hunk 7]
```diff
@@ -5,11 +5,16 @@
...
+using Libplanet.Common.Types.Assets;
+using Libplanet.Common.Types.Blocks;
+using Libplanet.Common.Types.Tx;
 using Serilog;
-using FAV = Libplanet.Assets.FungibleAssetValue;
+using FAV = Libplanet.Common.Types.Assets.FungibleAssetValue;
+using Transaction = Libplanet.Common.Types.Tx.Transaction;
+using TxExecution = Libplanet.Common.Types.Tx.TxExecution;
+using TxFailure = Libplanet.Common.Types.Tx.TxFailure;
+using TxSuccess = Libplanet.Common.Types.Tx.TxSuccess;
```
Comment (OnedgeLee) [inline Libplanet/Store/BaseStore.cs:17]: Do these classes can be ambiguous? Seems like needless `using`.

File: Libplanet/Blockchain/BlockChain.cs [hunk 8]
```diff
@@ -10,33 +10,36 @@
...
-    /// and pass it to the <see cref="BlockChain(IBlockPolicy, IStagePolicy,
-    /// IStore, IStateStore, Block, IBlockChainStates, IActionEvaluator, IEnumerable{IRenderer})"/>
+    /// and pass it to the
+    /// <see cref="BlockChain(IBlockPolicy, IStagePolicy, IStore, IStateStore, Block, IBlockChainStates, IActionEvaluator, IEnumerable{IRenderer})"/>
     /// constructor.
     /// </para>
     /// </summary>
     /// <remarks>This object is guaranteed that it has at least one block, since it takes a genesis
     /// block when it's instantiated.</remarks>
+ #pragma warning restore MEN002
```
[the remark this answers is not in the document]
Comment (riemannulus) [reply Libplanet/Blockchain/BlockChain.cs:42]: Cannot use pragma in XML documentation. 😢

[the remark this answers is not in the document]
Comment (OnedgeLee) [reply Libplanet/Blockchain/BlockChain.cs:42]: Meant, using `<see cref="BlockChain">'s constructor` would be enough, but may be not at second glance.

File: Docs/build.ps1 [hunk 9]
```diff
@@ -47,7 +47,7 @@ try {
 Set-Location $BaseDir
 dotnet tool run docfx docfx.json @args
 
-if (-not (Test-Path "$BaseDir/_site/api/Libplanet.html")) {
+if (-not (Test-Path "$BaseDir/_site/api/index.html")) {
```
Comment (moreal) [inline Docs/build.ps1:50]: Since the `Libplanet` namespace was gone, it should update `docs/api/index.md` like

```
<meta http-equiv="refresh" content="0; url=Libplanet.Common.html">
Move to [Libplanet.Common](Libplanet.Common.yml)&hellip;
```

And also `docs/toc.yml`

```#- name: Articles
#  href: articles/
- name: Overview
  href: articles/overview.md
- name: Design
  href: articles/design.md
- name: API Reference
  href: api/
  homepage: api/Libplanet.Common.yml
- name: Analyzer
  href: ../Libplanet.Analyzers/
  homepage: ../Libplanet.Analyzers/README.md
- name: CLI Tools
  href: ../Libplanet.Tools/README.md
- name: Explorer
  href: ../Libplanet.Explorer/
  homepage: ../Libplanet.Explorer/README.md
- name: Changelog
  href: ../CHANGES.md
- name: Contribute
  href: ../CONTRIBUTING.md
```

Comment (riemannulus) [reply Docs/build.ps1:50]: Can you make an issue about this?

Comment (moreal) [reply Docs/build.ps1:50]: I opened a new issue and I'll mark this conversation as resolved.

**Review by OnedgeLee — APPROVED** [review]

**Comment by longfin** [comment]
> > how about .Core? (or, is it already dropped candidate?
> 
> It was in candidates. And yes, Common is not good enough. This PR is just suggestion so we need to discussion about new project name.

(Just note, please correct it if I remember wrongly. 🙏 )

We've discussed `.Common` vs `.Core` and decided to postpone renaming to the next PR.
